diff -urN openssh-5.4p1/servconf.c /repo/openssh/595935/openssh/devel/openssh-5.5p1/servconf.c
--- openssh-5.4p1/servconf.c	2010-05-31 09:55:22.000000000 +0200
+++ /repo/openssh/595935/openssh/devel/openssh-5.5p1/servconf.c	2010-05-31 09:56:13.000000000 +0200
@@ -1271,7 +1270,17 @@
 		charptr = (opcode == sAuthorizedKeysFile) ?
 		    &options->authorized_keys_file :
 		    &options->authorized_keys_file2;
-		goto parse_filename;
+		arg = strdelim(&cp);
+		if (!arg || *arg == '\0')
+			fatal("%s line %d: missing file name.",
+			    filename, linenum);
+		if (*activep && *charptr == NULL) {
+			*charptr = tilde_expand_filename(arg, getuid());
+			/* increase optional counter */
+			if (intptr != NULL)
+				*intptr = *intptr + 1;
+		}
+		break;
 
 	case sClientAliveInterval:
 		intptr = &options->client_alive_interval;
